org.lflang/src/org/lflang/generator/c/CGenerator.xtend

```diff
@@ -2041,25 +1985,20 @@ class CGenerator extends GeneratorBase {
         // struct has a place to hold the data produced by this reactor's
         // reactions and a place to put pointers to data produced by
         // the contained reactors.
-        generateInteractingContainedReactors(reactor, federate, body, constructorCode, destructorCode);
+        generateInteractingContainedReactors(reactor, federate, body, constructorCode);
                 
         // Next, generate the fields needed for each reaction.
-        generateReactionAndTriggerStructs(body, decl, constructorCode, destructorCode, federate)
-        if (body.length > 0) {
-            code.pr('''
-                typedef struct {
-                    «body.toString»
-                } «selfType»;
-            ''')
-        } else {
-            // There are no fields for the self struct.
-            // C compilers complain about empty structs, so we generate a placeholder.
-            code.pr('''
-                typedef struct {
-                    bool hasContents;
-                } «selfType»;
-            ''')
-        }
+        generateReactionAndTriggerStructs(body, decl, constructorCode, federate)
+        
+        // The first field has to always be a pointer to the list of
+        // of allocated memory that must be freed when the reactor is freed.
+        // This means that the struct can be safely cast to self_base_t.
+        code.pr('''
+            typedef struct {
+                struct allocation_record_t *allocations;
```

Comment:
Can this directly be `struct self_base_t* base` instead of `struct allocation_record_t* allocations`? Maybe it would be more future-proof and allow for expansion of `self_base_t`.

Reply:
> Can this directly be `struct self_base_t* base` instead of `struct allocation_record_t* allocations`? Maybe it would be more future-proof and allow for expansion of `self_base_t`.

I don't understand this suggestion.  The typedef here is defining the self struct of a specific reactor class, which is different for every reactor class. The code here ensure that the first thing in that struct is a pointer to a list of allocations. Since this is the case for all self structs, then in C, you can cast the specific reactor class struct to `self_base_t` and then access the list of allocations.  This is what enables a generic destructor. It takes a specific self struct as an argument, casts it, and frees all the memory on the list.

Reply:
Yes, but couldn't the struct definition for a specific reactor class be:
```C
typedef struct {
    struct self_base_t* base;
    // Rest of the code
}
```

Reply:
Sorry if I'm being dense, but I still don't understand. Your struct would have to be cast to some struct type whose first element is `self_base_t* base`. What would we call that type? `self_base_base_t`?

Reply:
I think the struct that I am proposing can still be cast to `self_base_t` using the same `(self_base_t)bar` casting notation, but if we now add members to the definition of the `self_base_t` struct, these members will also be included in all new instances of `foo_type_t` as well. This is "simulating" inheritance in C.

A good example of this type of inheritance is explained here in the context of extending the base PyObject type in the C Python API to create new types: https://docs.python.org/3/extending/newtypes_tutorial.html

Reply:
Ah, got it!  The `*` is what threw me off. In fact, I don't think it would work with the `*`.

Reply:
I think we could make it work, but it would require an additional step to dereference the pointer and then cast it if I'm not mistaken. In any case, I think not using a pointer is the way to go.